# gViz
 A novel tool for the visualization of co-expression networks.
 
The quantity of microarray data available on the Internet has grown dramatically over the past years and now represents millions of Euros worth of underused information. One way to use this data is through co-expression analysis. To avoid a certain amount of bias, such data must often be analyzed at the genome scale, for example by network representation. The identification of co-expression networks is an important means to unravel gene to gene interactions and the underlying functional relationship between them. However, it is very difficult to explore and analyze a network of such dimensions. Several programs (Cytoscape, yEd) have already been developed for network analysis; however, to our knowledge, there are no available GraphML compatible programs.

We designed and developed gViz, a GraphML network visualization and exploration tool. gViz is built on clustering coefficient-based algorithms and is a novel tool to visualize and manipulate networks of co-expression interactions among a selection of probesets (each representing a single gene or transcript), based on a set of microarray co-expression data stored as an adjacency matrix.

 More informartion in the [publication](https://doi.org/10.1186/1756-0500-4-452).

## Citing gViz:
R. Helaers, E. Bareke, B. De Meulder, M. Pierre, S. Depiereux, N. Habra & E. Depiereux - gViz, a novel tool for the visualization of co-expression networks. BMC Research Notes 2011, 4:452

doi:10.1186/1756-0500-4-452
